### PR TITLE
get_* helpers for returning omega/sigma matrices and labeled theta values

### PR DESCRIPTION
**Pull request by barrettk**
# Proposed functionality:

### Single model summary
```r
sum <- MOD1 %>% model_summary()
```
```r
> sum %>% get_theta()
     THETA1      THETA2      THETA3      THETA4      THETA5 
  2.3412300  58.7268000 529.6200000  -0.0807956   4.1189400 


> sum %>% get_omega()
           OMEGA(1,1) OMEGA(1,2)
OMEGA(1,1)   0.104694   0.000000
OMEGA(2,1)   0.000000   0.159368


> sum %>% get_sigma()
           SIGMA(1,1)
SIGMA(1,1)          1
```

### With multiple model summaries
```r
mods <- list(MOD1, mod)
sums <- mods %>% model_summaries()
```
```r
> sums %>% get_theta()
[[1]]
     THETA1      THETA2      THETA3      THETA4      THETA5 
  2.3171600  54.6151000 462.5140000  -0.0820082   4.1795900 

[[2]]
     THETA1      THETA2      THETA3      THETA4      THETA5 
  2.3412300  58.7268000 529.6200000  -0.0807956   4.1189400 


> sums %>% get_omega()
[[1]]
           OMEGA(1,1) OMEGA(1,2)
OMEGA(1,1)  0.0985328   0.000000
OMEGA(2,1)  0.0000000   0.156825

[[2]]
           OMEGA(1,1) OMEGA(1,2)
OMEGA(1,1)   0.104694   0.000000
OMEGA(2,1)   0.000000   0.159368


> sums %>% get_sigma()
[[1]]
           SIGMA(1,1)
SIGMA(1,1)          1

[[2]]
           SIGMA(1,1)
SIGMA(1,1)          1
```

closes https://github.com/metrumresearchgroup/bbr/issues/442

### CODE REVIEW

**Comment by seth127** [comment]
I like this approach. 

Do we have any models (maybe in `complex`?) that have off-diagonals for either OMEGA or SIGMA. I just want to make sure we can test against that to make sure we're parsing the matrices correctly.

**Comment by kylebaron** [comment]
I think the names on the matrices should be refactored; it doesn't really make sense to have both the row and column index in the row name   ... the 2nd row is all OMEGA 2 and the 3rd column is all OMEGA 3.  

Something like

```r
[[1]]
           OMEGA(1)   OMEGA(2)
OMEGA(1)  0.0985328   0.000000
OMEGA(2)  0.0000000   0.156825
```

or make it easier to work with in R

```r
[[1]]
           OMEGA_1   OMEGA_2
OMEGA_1  0.0985328   0.000000
OMEGA_2  0.0000000   0.156825
```

**Comment by barrettk** [comment]
> I like this approach.
> 
> Do we have any models (maybe in `complex`?) that have off-diagonals for either OMEGA or SIGMA. I just want to make sure we can test against that to make sure we're parsing the matrices correctly.

It seems `example2_saemimp` is the only one we can rely on. `acop-fake-bayes` and `acop-iov` do have the necessary info, but do not have any covariance data. Its possible that `1001` and `acop-onlysim` would based on the ctl files, but they are missing relevant data in the summary calls and cant be used.

```r
> sums %>% get_omega()
[[1]]
           OMEGA(1,1) OMEGA(1,2)
OMEGA(1,1)  0.0985328   0.000000
OMEGA(2,1)  0.0000000   0.156825

[[2]]
             OMEGA(1,1)  OMEGA(1,2)  OMEGA(1,3)   OMEGA(1,4)
OMEGA(1,1)  0.010410500 0.000319542 0.001977200 -0.000189864
OMEGA(2,1)  0.000319542 0.008278020 0.000872168  0.001005720
OMEGA(3,1)  0.001977200 0.000872168 0.013139700  0.004040370
OMEGA(4,1) -0.000189864 0.001005720 0.004040370  0.010987300
```

File: tests/testthat/test-get-param.R
Comment (barrettk) [reply tests/testthat/test-get-param.R:75]: For reference:
```r
# matches values reported in parameter estimates to location in matrix
# does not check symmetry
check_matrix <- function(matrix, par_df, .type = "OMEGA"){
  for(name.i in unique(par_df$parameter_names)){
    # message("spot checking ", name.i)
    spot_chk <- par_df$estimate[par_df$parameter_names == name.i]
    # Get positions from labels
    matrix_pos <- gsub(.type, "",name.i)
    matrix_pos <- gsub("[()]", "", matrix_pos)
    matrix_pos <- strsplit(matrix_pos, ",")
    row.i <- matrix_pos[[1]][1] %>% as.numeric()
    col.i <- matrix_pos[[1]][2] %>% as.numeric()

    expect_identical(matrix[row.i,col.i], spot_chk)
  }
}
```

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:7]: Seems a bit odd to use "bbi_nonmem_summary" directly here and then the `NM_SUM_CLASS` indirection in the `check_model_object_list` right below.

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:11]: Given 1) `NM_SUM_CLASS` and 2) a `SL_CLASS <- "bbi_summary_list"` mapping exists, should `SL_CLASS` be used here?

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:48]: The `with` documentation recommends against using in the programming context:

```
Note:

     For _interactive_ use this is very effective and nice to read.
     For _programming_ however, i.e., in one's functions, more care is
     needed, and typically one should refrain from using ‘with()’, as,
     e.g., variables in ‘data’ may accidentally override local
     variables, see the reference.
```

(Same comment for `get_sigma`.)

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:115]: `seq_along` would be better for guarding the 0 case, which is a good to do even if it should be unreachable.

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:123]: Okay as is, but could be simplified (in my eyes) by dropping `paste0` and expanding format string to include `seq(...)` result.  Or, given this code base uses `glue` a lot, could even be `glue::glue("{.type}_{seq(...)}")`.

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:111]: Okay as is, but you can simplify the index access: `map(matrix_pos, 1)`.

**Review by kyleam — COMMENTED** [review]
Looks like a good start.  Here are comments from an initial read-through.  I plan to have at least one more round where I look more closely at the implementation and test the functions a bit.

File: R/get-param.R
Comment (barrettk) [reply R/get-param.R:7]: We have similar code [here](https://github.com/metrumresearchgroup/bbr/blob/main/R/test-threads.R#L180-L184), but I take your point. Its interesting we dont seem to have one for `"bbi_summary_list"`
```r
> NM_SUM_CLASS_LIST
[1] "bbi_nonmem_summary" "bbi_model"          "list"    
```

File: R/get-param.R
Comment (barrettk) [reply R/get-param.R:11]: Oh I take the above comment back, will make the change to `SL_CLASS`

File: R/get-param.R
Comment (barrettk) [reply R/get-param.R:12]: I suppose we should have an else statement that stops the function if any other method is piped in. I like the handling we did [here](https://github.com/metrumresearchgroup/bbr/blob/main/R/test-threads.R#L180-L185), so i'll probably just improve upon that.

File: R/get-param.R
Comment (barrettk) [reply R/get-param.R:58]: @kyleam Im now wondering if we want to export this other function, rather than making it internal. I currently have something like this:

image

KyleB was [suggesting](https://github.com/metrumresearchgroup/bbr/issues/442#issuecomment-976746226) something like this (though return everything)

Comment (kyleam) [reply R/get-param.R:69]: Hmm, but given it doesn't return everything, I'm not sure what functionality it adds to the interface, so my initial reaction is that it'd be better to leave it internal.  (And, when in doubt, I think internal is a good default because it's straightforward to expose something in a later release, but it's breaking to go in the other direction.)

Comment (barrettk) [reply R/get-param.R]: @kyleam I actually just refactored it again to be more in line with what Kyle was suggesting, so it now has some value, and returns an object similar to how `model_summaries()` does:

### Single summary
```r
> SUM1 %>% get_param()
$omega
          OMEGA_1  OMEGA_2
OMEGA_1 0.0985328 0.000000
OMEGA_2 0.0000000 0.156825

$sigma
        SIGMA_1
SIGMA_1       1

$theta
     THETA1      THETA2      THETA3      THETA4      THETA5 
  2.3171600  54.6151000 462.5140000  -0.0820082   4.1795900
```

### summary list
```r
> sums %>% get_param()
$omega
$omega[[1]]
          OMEGA_1  OMEGA_2
OMEGA_1 0.0985328 0.000000
OMEGA_2 0.0000000 0.156825

$omega[[2]]
             OMEGA_1     OMEGA_2     OMEGA_3      OMEGA_4
OMEGA_1  0.010410500 0.000319542 0.001977200 -0.000189864
OMEGA_2  0.000319542 0.008278020 0.000872168  0.001005720
OMEGA_3  0.001977200 0.000872168 0.013139700  0.004040370
OMEGA_4 -0.000189864 0.001005720 0.004040370  0.010987300


$sigma
$sigma[[1]]
        SIGMA_1
SIGMA_1       1

$sigma[[2]]
        SIGMA_1
SIGMA_1       1


$theta
$theta[[1]]
     THETA1      THETA2      THETA3      THETA4      THETA5 
  2.3171600  54.6151000 462.5140000  -0.0820082   4.1795900 

$theta[[2]]
    THETA1     THETA2     THETA3     THETA4     THETA5     THETA6     THETA7     THETA8     THETA9    THETA10    THETA11 
 3.3005000  3.2519700 -0.6122190 -0.2080120  0.7372940  1.1359900  0.3346290  0.1923520  0.6916970  2.2993700  0.0991684 
```

Comment (kyleam) [reply R/get-param.R]: > so it now has some value

That's true, though I'm not really a fan.  I think we should choose one approach, not expose a function that returns all of the parameters _and_ accessor functions for the individual parameters.  It complicates the interface and code.  

That's of course just one opinion.  Perhaps others disagree with me here.

Comment (barrettk) [reply R/get-param.R]: The main thing I like is being _**able to**_ retrieve all parameters, but not necessarily forced to. Perhaps I could make `"omega"` the default instead of all 3. If we chose to keep the 3 specific functions (and hide `get_param()`, I think I would still leave the ability to map across all 3 parameters just in case we want that functionality in the future.

@callistosp @kylebaron what do you guys think of the above change (see code snippets)? Does it make sense to have 1 general function, a 3 specific functions in terms of how you would/could use them?

Comment (seth127) [reply R/get-param.R]: I think I agree with @kyleam to err on the side of keeping it simple here (i.e. _don't_ add the extra helper). Also, I think Sam already weighed in on this [on the issue](https://github.com/metrumresearchgroup/bbr/issues/442#issuecomment-1180619733) and seemed to agree that the helper wouldn't likely be necessary in most cases.

**Comment by barrettk** [comment]
@seth127 wondering if I could do this:

```r
#' @section Examples:
#'
#' ```{r, include = FALSE}
#' knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
#' options('bbr.bbi_exe_path' = read_bbi_path())
#' MODEL_DIR <- system.file("model", "nonmem", "complex",   package = "bbr")
#' mod <- read_model(file.path(MODEL_DIR, "example2_saemimp"))
#' ```
#'
#' ```{r, comment = "#>", collapse = TRUE}
#' sum <- mod %>% model_summary()
#'
#' sum %>% get_omega()
#'
#'
#' sum %>% get_sigma()
#'
#'
#' sum %>% get_theta()
#' ```
```

I think this will appease drone, but im worried about how this helper doc would render after someone installed the package on their end (without running `use_bbi()`). I would add hidden conditional logic that would use that function if not found at that path, but I have a feeling there are reasons we wouldnt want to do that. A lot of this package doesn't have examples (which I think would be nice), but I can see the difficulty in making them for most functions without having bbi installed. Have thought about this for a while, but wondering what your thoughts are (in this case, and having examples relying on bbi in general)

**Comment by seth127** [comment]
Good thoughts there @barrettk . At this point, I think our best move is to just do examples and make them `\dontrun{`. The main downside of that is that CRAN doesn't like it, but getting `bbr` on CRAN really isn't a priority because pretty much all likely users are gonna get it from MPN anyway. We already use this pattern in other places and you're right that we should probably add more examples for other functions.

> maybe having some version of bbi installed in inst or something (Im sure thats a terrible idea)

Definitely _not_ a terrible idea. I've thought about this a few times, not `inst` specifically, but somehow having a version of `bbi` included in `bbr`. Every time I run through it in my head, I end up deciding it's not a good idea, but I do keep coming back to it. There are a couple big advantages, a couple off the top:

* streamline getting started for new users
* `bbi` version pinned (to something that's compatible) by default and users only change it if they have a reason to.
* would make it easier for regulators or whoever else that wants to re-run code without having to work hard to recreate the environment.
* the test suite and example stuff that you pointed out

Anyway, something to consider at some point, though I doubt we'll bite it off anytime soon. Lemme know if you want to talk further at some point.

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:86]: This comes triggers a note:

```
❯ checking Rd line widths ... NOTE
  Rd file 'get_omega.Rd':
    \examples lines wider than 100 characters:
           THETA1     THETA2     THETA3     THETA4     THETA5     THETA6     THETA7     THETA8     THETA9    THETA10    THETA11
        3.3005000  3.2519700 -0.6122190 -0.2080120  0.7372940  1.1359900  0.3346290  0.1923520  0.6916970  2.2993700  0.0991684
```

That's not really a problem, though it is nice to keep the check output as quiet as possible.  Maybe it'd be enough to just show, say, the `get_omega()` calls.  An example is nice, but once you've seen one, I'm not sure seeing each other one adds much value.  Up to you, just a thought.

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:25]: Missing `map` `@importFrom` for this function.

File: tests/testthat/test-get-param.R
Comment (kyleam) [inline tests/testthat/test-get-param.R:9]: Stale comment as of 620b1064?

File: tests/testthat/test-get-param.R
Comment (kyleam) [inline tests/testthat/test-get-param.R:70]: Typo:  "coviarances" (here and in other spots)

File: R/get-param.R
Comment (kyleam) [inline R/get-param.R:13]: I'm confused by `check_model_object_list` is achieving here.  In the condition right above, you just did the same `inherits` check on `NM_SUM_CLASS` that `check_model_object_list` did underneath.  So isn't the failure unreachable?

https://github.com/metrumresearchgroup/bbr/blob/033899436217eb6a24083f97ae2063933b75a828/R/utils.R#L385-L386

Similar comment for the `check_model_object` call for the next condition.

**Review by kyleam — COMMENTED** [review]
Thanks for the update.  A few more minor comments, but looking almost ready to me.

File: R/get-param.R
Comment (barrettk) [reply R/get-param.R:13]: Yes I had this thought after merging the test_threads PR a while back (used in `check_run_times()`). I forget how I had it originally, but Seth and I came up with that together. I dont think either of us fully inspected what the `check_` functions were doing at that point (we probably assumed it did more than confirm the class). In the case of `check_run_times()` we wanted to ensure a model object, model list, model summary, or model summary list object was passed in, and in this case we only want to allow a model summary or model summary list. Because of classes we weren't able to use solely `check_model_object` or `check_model_object_list`.



Ill rework this to make it actually meaningful. Will likely ditch the check_ functions altogether

Comment (kyleam) [reply R/get-param.R:13]: Okay, to be clear, not a big deal, I was just scratching my head.

Comment (barrettk) [reply R/get-param.R]: No worries. Made a change that I think improves the logic. Will hold off on touching the other function.

**Review by kyleam — APPROVED** [review]
Update looks good to me.  Thanks.

**Pull request merged**
